### PR TITLE
Bulk Export Bug With Many Resources and Low Max File Size

### PR DESCRIPTION
**Pull request by nathandoef**
Previously, bulk export jobs were getting stuck in the `FINALIZE` state when performed with many resources and a low Bulk Export File Maximum Capacity. This has been fixed.

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java [hunk 1]
```diff
@@ -325,12 +321,17 @@ public Iterator<WorkChunk> fetchAllWorkChunksIterator(String theInstanceId, bool
 		return new PagingIterator<>((thePageIndex, theBatchSize, theConsumer) -> fetchChunks(theInstanceId, theWithData, theBatchSize, thePageIndex, theConsumer));
 	}
 
-	/**
-	 * Note: Not @Transactional because the transaction happens in a lambda that's called outside of this method's scope
-	 */
 	@Override
+	@Transactional(propagation = Propagation.REQUIRES_NEW)
 	public Iterator<WorkChunk> fetchAllWorkChunksForStepIterator(String theInstanceId, String theStepId) {
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java:326]: Throwing the contents of the stream into a List risks blowing out memory.  Can we change fetchAllWorkChunksForStepIterator to simply return the stream instead?

Comment (nathandoef) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java:335]: Good call, changes made

**Review by JasonRoberts-smile — APPROVED** [review]
Looks good. I like the use of streams

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/ReductionStepExecutor.java [hunk 2]
```diff
@@ -67,97 +73,105 @@ <PT extends IModelJson, IT extends IModelJson, OT extends IModelJson> boolean ex
...
+		} else {
+			try {
+				// feed them into our reduction worker
+				// this is the most likely area to throw,
+				// as this is where db actions and processing is likely to happen
+				ChunkExecutionDetails<PT, IT> chunkDetails = new ChunkExecutionDetails<>(theChunk.getData(theInputType), theParameters, theInstance.getInstanceId(), theChunk.getId());
+
+				ChunkOutcome outcome = theReductionStepWorker.consume(chunkDetails);
+
+				switch (outcome.getStatuss()) {
```
[the remark this answers is not in the document]
Comment (nathandoef) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/ReductionStepExecutor.java]: Good catch, fixed

**Pull request merged**